cdap-docs/developers-manual/source/advanced/class-loading.rst

```diff
@@ -0,0 +1,100 @@
+
+Moreover, CDAP allows an application to define its own plugin API for extending the
+functionalities of the application itself. Each plugin implementation will be loaded
+through a separate plugin class loader. The parent of each plugin class loader is a filter
+class loader that ensures that only CDAP and plugin API classes are loaded from the
+application class loader. All other classes required by the plugin are loaded by the
+plugin class loader itself. This allows an application to use multiple plugins at the same
+time, while providing class loading isolation between different plugins.
+
+In CDAP, both applications and plugins are deployed as JAR files to the system. Instead of
+a normal JAR file that contains a collection of class files, a bundle JAR file has content
+similar to this example:
+
+.. image:: ../_images/class-loading/class-loading02.png
+   :align: center
+
+..    :width: 58%
+..    :scale: 75%
+
+Inside an application bundle JAR file, besides the class files for the application, are
+the JAR files for third party libraries that are needed by the application. In the
+application JAR manifest, a special attribute (“Export-Package”) can be added to instruct
+the filter class loader which packages are contained in the application plugin API. Hence,
+only classes under those packages will be available to the plugin. This is to avoid
+libraries used by the application interfering with the ones used by the plugin. The plugin
+JAR file is following a similar structure as an application JAR file.
+
+Class Loading in Apache Hadoop MapReduce and Apache Spark
+=========================================================
+
+One of the major challenges faced when we integrate CDAP with data processing frameworks,
+such as MapReduce and Spark, is the class loading. Both frameworks use a
+flat classpath approach. For example, in MapReduce, all the job classes, the
+libraries that the job needs, together with the Hadoop libraries and their dependencies,
+are in one classpath. The classpath is used to create a single class loader, which loads
+classes by sequentially searching through the classpath. This approach doesn’t provide the
+class loading isolation that is desired. For example, if a MapReduce job requires a
+different version of a library than the one used by the Hadoop framework in the cluster,
+the job can fail with exceptions such as ``NoSuchMethodError`` or ``ClassNotFoundException``.
+
+When launching a MapReduce or Spark job from CDAP, we want to provide the same class
+isolation behavior as described in the previous section. To do so, we alter the classpath
+of the YARN container for the job and make it first call the main method of our launcher
+class. The launcher class will setup the class loader hierarchy as desired and then
+delegate the call to the actual MapReduce or Spark task runner.
+
+Summary
+=======
+
+Java class loading is a powerful and important concept provided by the Java platform.
```

Comment:
I don't think we need this summary. This is juts a chapter in the dev guide, not a blog or article that strands by itself.